Cifti matlab settings and rearrangement

coalsont opened the pull request:
The rearrangement has not been tested at all, do not merge until tested.

glasserm commented on ICAFIX/PostFix.sh:
```diff
@@ -409,7 +409,7 @@ main()
 	# For interpreted modes, make sure that matlab/octave have access to the functions they need
 	export FSL_MATLAB_PATH="${FSLDIR}/etc/matlab"
-	local ML_PATHS="addpath('${FSL_MATLAB_PATH}'); addpath('${MATLAB_GIFTI_LIB}'); addpath('${this_script_dir}/scripts');"
+	local ML_PATHS="addpath('${FSL_MATLAB_PATH}'); addpath('$HCPCIFTIRWDIR'); addpath('${MATLAB_GIFTI_LIB}'); addpath('${this_script_dir}/scripts');"
```
Should probably name the MATLAB_GIFTI_LIB variable consistently with other functions.

coalsont replied:
The other scripts in ICAFIX/ use `'${HCPPIPEDIR}/global/matlab'`, not a variable.  The folder does contain the gifti library, so it isn't entirely inaccurate.  What are you suggesting, specifically?

glasserm replied:
I think Tim brown wrote that before he understood as much.  That code does not use gifti, it is what we reference in the other scripts as '${HCPPIPEDIR}/global/matlab'.

glasserm commented:
Lets call the new folder 'cifti-legacy'

mharms commented on ICAFIX/ReApplyFixMultiRunPipeline.sh:
```diff
@@ -501,10 +501,10 @@ main()
-	# BUT the CIFTI I/O functions are also now included within '${HCPPIPEDIR}/global/matlab' as well.
-	# AND since 'addpath' *prepends* to the matlab path, the versions in '${HCPPIPEDIR}/global/matlab' will therefore take precedence!
+	# The pipelines CIFTI I/O functions have been moved to a subdirectory, and the SetUp... script selects what to use with HCPCIFTIRWDIR.
+	# AND since 'addpath' *prepends* to the matlab path, the versions in '${HCPPIPEDIR}/global/matlab' will therefore take precedence.
```
Comment should be changed to:
```suggestion
	# AND since 'addpath' *prepends* to the matlab path, the versions in '${HCPCIFTIRWDIR}' will therefore take precedence.
```

glasserm replied:
well it all depends on the order of the addpath commands...

mharms replied:
Well, `${HCPPIPEDIR}/global/matlab` will have precedence in the path, but it won't contain any CIFTI I/O functions, which is what is being referred to here.  But, I see the potential source of confusion regarding my suggestion, and have edited it appropriately.

coalsont replied:
Since FSL_FIX_CIFTIRW isn't mentioned in the file, it would probably be better to say something like "take precedence over the fix settings file".

mharms replied:
That's fine with me for the line in question, although given the convoluted nature by which this all works together, I think it is still valuable to maintain the mentions of  `${FSL_FIX_CIFTIRW}` and `${FSL_FIX_WBC}` earlier in the comment.

mharms commented on ICAFIX/ReApplyFixPipeline.sh:
```diff
@@ -395,8 +395,8 @@ main()
 	# environment variable within fix_3_clean.m itself.
-	# Note that we are NOT adding '${HCPPIPEDIR}/global/matlab' to the path (which also contains the
-	# CIFTI I/O functions as well), so the versions in ${FSL_FIX_CIFTIRW} will be the ones that get used
+	# Note that we are NOT adding '${HCPPIPEDIR}/global/matlab' to the path (which used to contain the
+	# CIFTI I/O functions as well), or HCPCIFTIRWDIR, so the versions in ${FSL_FIX_CIFTIRW} will be the ones that get used
```
Should we consider changing this behavior at this time?  i.e., Make the CIFTI I/O in this script use what is specified by `${HCPCIFTIRWDIR}` rather than `${FSL_FIX_CIFTIRW}` (for interpreted mode)?

glasserm replied:
I think it would be reasonable to override anything in the FIX settings.

mharms commented on ICAFIX/scripts/Compile_MATLAB_code.sh:
```diff
@@ -72,7 +59,7 @@ compile_prepareICAs()
 	log_Msg "Compiling ${app_name} application"
 	${MATLAB_HOME}/bin/mcc -m -v ${app_name}.m \
-				  -I ${HCPPIPEDIR}/global/matlab/gifti-1.6 \
+				  -I ${HCPCIFTIRWDIR} \
 				  -I ${HCPPIPEDIR}/global/matlab \
```
Now that we've eliminated the `@gifti` symlink, we could make the path order for the compilation consistent with the order that we are using in setting the path (via `addpath`) for the interpreted matlab.  That order would be:
```
...
-I ${HCPPIPEDIR}/global/matlab \
-I ${HCPCIFTIRWDIR} \
...
```
i.e., give `${HCPPIPEDIR}/global/matlab` precedence (given that -I appends).

mharms commented on MSMAll/scripts/MSMAll.sh:
```diff
@@ -722,7 +722,7 @@ main()
 						mGlobalPath="${HCPPIPEDIR}/global/matlab"
 
-						matlabCode="addpath '$mPath'; addpath '$mGlobalPath'; MSMregression('${inputspatialmaps}','${inputdtseries}','${inputweights}','${outputspatialmaps}','${outputweights}','${Caret7_Command}','${Method}','${Params}','${VN}',${nTPsForSpectra},'${BC}','${VolParams}');"
+						matlabCode="addpath '$mPath'; addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath'; MSMregression('${inputspatialmaps}','${inputdtseries}','${inputweights}','${outputspatialmaps}','${outputweights}','${Caret7_Command}','${Method}','${Params}','${VN}',${nTPsForSpectra},'${BC}','${VolParams}');"
```
If we want a consistent principle on path precedence, we would do the following instead:
```suggestion
						matlabCode="addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath'; addpath '$mPath';  MSMregression('${inputspatialmaps}','${inputdtseries}','${inputweights}','${outputspatialmaps}','${outputweights}','${Caret7_Command}','${Method}','${Params}','${VN}',${nTPsForSpectra},'${BC}','${VolParams}');"
```

mharms commented on MSMAll/scripts/MSMAll.sh:
```diff
@@ -807,7 +807,7 @@ main()
 						mGlobalPath="${HCPPIPEDIR}/global/matlab"
 						
-						matlabCode="addpath '$mPath'; addpath '$mGlobalPath';
+						matlabCode="addpath '$mPath'; addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath';
```
```suggestion
						matlabCode="addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath'; addpath '$mPath'; 
```

mharms commented on MSMAll/scripts/MSMAll.sh:
```diff
@@ -1225,7 +1225,7 @@ main()
 				mGlobalPath="${HCPPIPEDIR}/global/matlab"
 				
-				matlabCode="addpath '$mPath'; addpath '$mGlobalPath';
+				matlabCode="addpath '$mPath'; addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath';
```
```suggestion
				matlabCode="addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath'; addpath '$mPath'; 
```

mharms commented on MSMAll/scripts/MSMAll.sh:
```diff
@@ -1294,7 +1294,7 @@ main()
 				mGlobalPath="${HCPPIPEDIR}/global/matlab"
 				
-				matlabCode="addpath '$mPath'; addpath '$mGlobalPath';
+				matlabCode="addpath '$mPath'; addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath';
```
```suggestion
				matlabCode="addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath'; addpath '$mPath'; 
```

mharms commented on MSMAll/scripts/SingleSubjectConcat.sh:
```diff
@@ -421,7 +421,7 @@ main()
 						mGlobalPath="${HCPPIPEDIR}/global/matlab"
 						
-						matlabCode="addpath '$mPath'; addpath '$mGlobalPath';
+						matlabCode="addpath '$mPath'; addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath';
```
```suggestion
						matlabCode="addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath'; addpath '$mPath'; 
```

mharms commented on RestingStateStats/RestingStateStats.sh:
```diff
@@ -968,7 +968,7 @@ main()
 			mFslPath="${FSLDIR}/etc/matlab"
 
-			matlabCode="addpath '$mPath'; addpath '$mGlobalPath'; addpath '$mFslPath'; RestingStateStats('${motionparameters}',${g_high_pass},${TR},'${ICAs}','${noise}','${CARET7DIR}/wb_command','${dtseries}','${bias}','${RssPrefix}','${g_dlabel_file}','${g_bc_mode}','${g_out_string}','${WM}','${CSF}');"
+			matlabCode="addpath '$mPath'; addpath '$HCPCIFTIRWDIR' addpath '$mGlobalPath'; addpath '$mFslPath'; RestingStateStats('${motionparameters}',${g_high_pass},${TR},'${ICAs}','${noise}','${CARET7DIR}/wb_command','${dtseries}','${bias}','${RssPrefix}','${g_dlabel_file}','${g_bc_mode}','${g_out_string}','${WM}','${CSF}');"
```
In addition to path order consistency, existing command missing a semicolon between two `addpath`
```suggestion
			matlabCode="addpath '$mFslPath'; addpath '$HCPCIFTIRWDIR'; addpath '$mGlobalPath'; addpath '$mPath'; RestingStateStats('${motionparameters}',${g_high_pass},${TR},'${ICAs}','${noise}','${CARET7DIR}/wb_command','${dtseries}','${bias}','${RssPrefix}','${g_dlabel_file}','${g_bc_mode}','${g_out_string}','${WM}','${CSF}');"
```

mharms commented on global/matlab/Compile_All_MATLAB_code.sh:
```diff
@@ -113,50 +113,28 @@ get_options()
-# ------------------------------------------------------------------------------
-#  Main processing of script.
-# ------------------------------------------------------------------------------
-main()
-{
-	get_options "$@"
-	
-	compile_ICAFIX_code
-	compile_MSMALL_code
-	compile_RestingStateStats_code
-}
```
I would prefer that we not make these changes to the "master" compile script. Having three separate `compile_*_code` functions within `main` makes it very convenient to comment out the sub-pipelines that don't need re-compiling (e.g., if we only changed ICAFIX code, and thus only need to recompile that).

coalsont replied:
1) Scripts already exist to compile only a single pipeline's matlab code, and require no editing, and are co-located with that pipeline's matlab files, which should be more convenient than editing anything.
2) A function that calls only one line of processing code needlessly lengthens a script and makes it look more complicated.
3) You can still comment out only one line to prevent the actual compilation step, and I would be perfectly happy removing the `echo`s entirely, but I have never had to read the script output.
4) If the script should officially support not compiling everything, then that suggests it should have options for this, rather than requiring editing.

Given that the only option actually supported is the help info option, and the scripts it calls do their own checking of environment variables anyway, I would instead be tempted to replace the whole thing with a 5-line script:
```
#!/bin/bash
set -eux
"$HCPPIPEDIR"/ICAFIX/scripts/Compile_MATLAB_code.sh
"$HCPPIPEDIR"/MSMAll/scripts/Compile_MATLAB_code.sh
"$HCPPIPEDIR"/RestingStateStats/scripts/Compile_MATLAB_code.sh
```

mharms replied:
I see your points, regarding the originally proposed changes.  I'm not a fan though of going with the entirely minimalist approach (just 3 lines), as I think the Usage statement adds valuable context and information.

mharms commented on global/scripts/RSNregression.sh:
```diff
@@ -92,7 +92,7 @@ case "$MatlabMode" in
         ;;
 esac
-matlab_paths="addpath('$FSLDIR/etc/matlab'); addpath('$HCPPIPEDIR/global/matlab'); addpath('$this_script_dir');
+matlab_paths="addpath('$FSLDIR/etc/matlab'); addpath('$HCPPIPEDIR/global/matlab'); addpath('$HCPCIFTIRWDIR'); addpath('$this_script_dir');
```
```suggestion
matlab_paths="addpath('$FSLDIR/etc/matlab'); addpath('$HCPCIFTIRWDIR'); addpath('$HCPPIPEDIR/global/matlab'); addpath('$this_script_dir');
```

mharms commented on ICAFIX/ReApplyFixPipeline.sh:
```diff
@@ -395,11 +395,12 @@ main()
-	# (assuming that the user hasn't further customized their matlab path, e.g., via a startup.m addition).
+	# Note that previously we did not add '${HCPPIPEDIR}/global/matlab' to the path (which used to contain the
+	# CIFTI I/O functions as well), or HCPCIFTIRWDIR.
+	# Since addpath adds to the front of the matlab path, we are now overriding the fix settings for cifti IO,
```
```suggestion
	# Since addpath adds to the front of the matlab path, we are now overriding the fix settings for cifti I/O,
```

mharms commented on ICAFIX/scripts/Compile_MATLAB_code.sh:
```diff
@@ -33,26 +33,13 @@
+# not only does not follow them, it also remembers that they exist and refuses to allow anything with the same
+# name found in a later -I option to be used.  @gifti used to be a symlink, requiring the compilation to
+# include the real location of the @gifti folder, before a folder that contains an @gifti symlink.
+
```
Restoring this tidbit, so that the behavior of -I is explicit, and I don't need to translate the implied ordering when I return to this script in the future
```suggestion
# (since the -I option *appends* folders to the search path, in the order listed)

```

coalsont commented:
I haven't worked out a plan for testing yet.  Maybe we can wait until we are ready to make cifti-matlab the default, and test it then?

glasserm commented:
I thought the point of this was to enable testing of CIFTI matlab.  RSNRegression works.

coalsont commented:
It also moves some things around to more permanently resolve some hazards in the prior matlab compiler setup, and edits several additional scripts that use cifti IO.  I guess if we feel the edits are sufficiently similar, we don't have to test all of them.

We could extend this PR to also be where we add a copy of cifti-matlab (once it has stabilized somewhat), since that is what it is preparing for anyway.  Of course, if you want to be testing cifti-matlab against master for a while before we do that, that means we should prioritize merging this and not extend it.

mharms commented:
My main concern about timing is the potential for merge conflicts to develop, although I don't foresee that being a big problem...

coalsont commented:
It may be time to do a non-beta release of cifti-matlab and finish this PR.

glasserm commented:
I would be fine with a CIFTI matlab release.

coalsont commented:
I have added cifti-matlab and set it as the default in `Examples/Scripts/SetUp...` and modified the LICENSE.md to note the GPL from some of the code (including gifti-1.6 we were already using).  I am talking to the relevant authors about dual-licensing, and will update the licenses depending on the outcome.

mharms commented:
What are your thoughts about the compiled matlab?  Do we recompile so that the compiled is based on the same `cifti-matlab` library?

glasserm commented:
Yes we should do that before the next release.

coalsont commented:
Because the gifti toolbox is staying GPL (derived from SPM), and no released scripts use it (it is just a dependency of `cifti-legacy`), I have moved it inside of cifti-legacy, so that inclusion of `global/matlab` doesn't automatically drag it in (and using `cifti-legacy` doesn't require including `global/matlab` anymore).  If in-progress scripts use `@gifti` directly, perhaps the thing to do is make a second copy in the original `gifti-1.6` folder (and not make a symlink, so the matlab compiler doesn't get confused).

`@xmltree` is now LGPL v3+ and `cifti-matlab` is BSD, so `cifti-matlab` can be compiled together with whatever other code under whatever licenses.  Additionally, `fsl/etc/matlab/read_avw.m`, etc have been released under BSD terms per Steve Smith (via email), so it won't be a problem to compile with GPL code anyway.

glasserm commented:
I don't think we load GIFTI files in matlab often and with the new toolbox I expect that to be even less.

coalsont commented:
We should probably do a quick test on the current branch of something that uses cifti, then I think this is finally ready for merge.

glasserm commented:
I think we should finish this up.  I have been on this branch in my version of the repo for quite some time and would like to switch back to master.
